### PR TITLE
Support unparsing plans after applying `optimize_projections` rule 

### PR DESCRIPTION
**Pull request by sgrebnov**
## Which issue does this PR close?

The `optimize_projections` optimization is very useful as it pushes down projections to the `TableScan` and ensures only required columns are fetched. This is useful when used alongside unparsing scenarios for plans involving multiple data sources, as the plan must be optimized to push down projections and fetch only the required columns. The downside of this process is that the rule modifies the original plan in a way that makes it difficult to unparse, and the resultant plan is not always optimal or efficient for unparsing use cases, for example

https://gist.github.com/sgrebnov/5071d2834e812b62bfdf434cf7e7e54c

Original query (TPC-DS Q72)
```sql
select  i_item_desc
      ,w_warehouse_name
      ,d1.d_week_seq
      ,sum(case when p_promo_sk is null then 1 else 0 end) no_promo
      ,sum(case when p_promo_sk is not null then 1 else 0 end) promo
      ,count(*) total_cnt
from catalog_sales
join inventory on (cs_item_sk = inv_item_sk)
join warehouse on (w_warehouse_sk=inv_warehouse_sk)
join item on (i_item_sk = cs_item_sk)
join customer_demographics on (cs_bill_cdemo_sk = cd_demo_sk)
join household_demographics on (cs_bill_hdemo_sk = hd_demo_sk)
join date_dim d1 on (cs_sold_date_sk = d1.d_date_sk)
join date_dim d2 on (inv_date_sk = d2.d_date_sk)
join date_dim d3 on (cs_ship_date_sk = d3.d_date_sk)
left outer join promotion on (cs_promo_sk=p_promo_sk)
left outer join catalog_returns on (cr_item_sk = cs_item_sk and cr_order_number = cs_order_number)
where d1.d_week_seq = d2.d_week_seq
  and inv_quantity_on_hand < cs_quantity
  and d3.d_date > d1.d_date + INTERVAL '5 days'
  and hd_buy_potential = '501-1000'
  and d1.d_year = 1999
  and cd_marital_status = 'S'
group by i_item_desc,w_warehouse_name,d1.d_week_seq
order by total_cnt desc, i_item_desc, w_warehouse_name, d_week_seq
 LIMIT 100;
```

Plan and query after applying `optimize_projections` rule. Notice the additional projections added after joins.

image

```sql
select
	"i_item_desc",
	"w_warehouse_name",
	"d_week_seq",
	sum(case when "p_promo_sk" is null then 1 else 0 end) as "no_promo",
	sum(case when "p_promo_sk" is not null then 1 else 0 end) as "promo",
	count(1) as "total_cnt"
from
	(
	select
		"w_warehouse_name",
		"i_item_desc",
		"d_week_seq",
		"p_promo_sk"
	from
		(
		select
			"cs_item_sk",
			"cs_order_number",
			"w_warehouse_name",
			"i_item_desc",
			"d_week_seq",
			"promotion"."p_promo_sk"
		from
			(
			select
				"cs_item_sk",
				"cs_promo_sk",
				"cs_order_number",
				"w_warehouse_name",
				"i_item_desc",
				"d_week_seq"
			from
				(
				select
					"cs_ship_date_sk",
					"cs_item_sk",
...
```

## Rationale for this change

To support unparsing plans after `optimize_projections` is applied, it is proposed to add the `optimize_projections_preserve_existing_projections` configuration option to prevent the optimization logic from creating or removing projections and to preserve the original structure. It ensures the query layout remains simple and readable, relying on the underlying SQL engine to apply its own optimizations during execution.

## Are these changes tested?

Added test for `optimize_projections_preserve_existing_projections` configuration option. Unparsing have been tested by running all TPC-H and TPC-DS queries with `optimization_projections` enabled.

## Are there any user-facing changes?

Yes, a new `optimize_projections_preserve_existing_projections` configuration option has been introduced, which can be specified via `SessionConfig` or at a lower level using `OptimizerContext::new_with_options`.

```rust
SessionStateBuilder::new()
  .with_config(
      SessionConfig::new().with_optimize_projections_preserve_existing_projections(true),
  )
  .build();
```
Or 

```rust
let mut config = ConfigOptions::new();
config
    .optimizer
    .optimize_projections_preserve_existing_projections =
    preserve_projections;
let optimizer_context = OptimizerContext::new_with_options(config);
```

There are no changes in default behavior.

### CODE REVIEW

File: datafusion/sql/src/unparser/plan.rs [hunk 1]
```diff
@@ -725,24 +727,29 @@ impl Unparser<'_> {
...
-                                    field.name().clone(),
-                                )
-                            }
-                        })
-                        .collect::<Vec<_>>();
-                    builder = builder.project(project_columns)?;
+                // Avoid creating a duplicate Projection node, which would result in an additional subquery if a projection already exists.
+                // For example, if the `optimize_projection` rule is applied, there will be a Projection node, and duplicate projection 
+                // information included in the TableScan node.
+                if !already_projected {
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/plan.rs:733]: This prevents from generating queries like `SELECT a, b from (SELECT a, b from my_table)`.

File: datafusion/sql/tests/cases/plan_to_sql.rs [hunk 2]
```diff
@@ -882,6 +882,7 @@ fn test_table_scan_pushdown() -> Result<()> {
     let query_from_table_scan_with_projection = LogicalPlanBuilder::from(
         table_scan(Some("t1"), &schema, Some(vec![0, 1]))?.build()?,
     )
+    .project(vec![col("id"), col("age")])?
     .project(vec![wildcard()])?
```
Comment (sgrebnov) [inline datafusion/sql/tests/cases/plan_to_sql.rs:886]: Can this actually be a real plan with a wildcard projection and a TableScan that includes only two columns? I would expect them to match. If this is a real use case I will improve logic above to not duplicate projection (check for parent projection is a wildcard or does not match). Running all TPC-H and TPC-DS queries I've not found query where it was the case.
https://github.com/apache/datafusion/pull/13267/files#r1830100022

**Comment by findepi** [comment]
Would this be equivalent to disabling `optimize_projections`?

**Comment by sgrebnov** [comment]
> Would this be equivalent to disabling `optimize_projections`?

@findepi  -  the logic of `optimize_projections` is still working in this case pushing down only required columns, for example

```sql
explain SELECT ss_item_sk, trip_distance_mi, total_amount
FROM store_sales
JOIN taxi_trips
ON store_sales.ss_quantity = taxi_trips.passenger_count
WHERE taxi_trips.total_amount=3.0 and ss_item_sk= 180
ORDER BY trip_distance_mi DESC LIMIT 10;
```

Without `optimize_projections` 

```console
| logical_plan  | Sort: taxi_trips.trip_distance_mi DESC NULLS FIRST, fetch=10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |   Projection: store_sales.ss_item_sk, taxi_trips.trip_distance_mi, taxi_trips.total_amount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |     Inner Join: CAST(store_sales.ss_quantity AS Int64) = taxi_trips.passenger_count                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |       Projection: store_sales.ss_item_sk, store_sales.ss_quantity                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |         BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |           Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |  Projection: store_sales.ss_sold_date_sk, store_sales.ss_sold_time_sk, store_sales.ss_item_sk, store_sales.ss_customer_sk, store_sales.ss_cdemo_sk, store_sales.ss_hdemo_sk, store_sales.ss_addr_sk, store_sales.ss_store_sk, store_sales.ss_promo_sk, store_sales.ss_ticket_number, store_sales.ss_quantity, store_sales.ss_wholesale_cost, store_sales.ss_list_price, store_sales.ss_sales_price, store_sales.ss_ext_discount_amt, store_sales.ss_ext_sales_price, store_sales.ss_ext_wholesale_cost, store_sales.ss_ext_list_price, store_sales.ss_ext_tax, store_sales.ss_coupon_amt, store_sales.ss_net_paid, store_sales.ss_net_paid_inc_tax, store_sales.ss_net_profit                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
|               |   Filter: CAST(store_sales.ss_item_sk AS Int64) = Int64(180)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |     TableScan: store_sales, partial_filters=[CAST(store_sales.ss_item_sk AS Int64) = Int64(180)]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |       Projection: taxi_trips.passenger_count, taxi_trips.trip_distance_mi, taxi_trips.total_amount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |         BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |           Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |  Projection: taxi_trips.pickup_datetime, taxi_trips.passenger_count, taxi_trips.trip_distance_mi, taxi_trips.fare_amount, taxi_trips.tip_amount, taxi_trips.total_amount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
|               |   TableScan: taxi_trips, full_filters=[taxi_trips.total_amount = Float64(3)]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
```

With `optimize_projections`  and `optimize_projections_preserve_existing_projections `

```console
| logical_plan  | Sort: taxi_trips.trip_distance_mi DESC NULLS FIRST, fetch=10                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |   Projection: store_sales.ss_item_sk, taxi_trips.trip_distance_mi, taxi_trips.total_amount                                                                                                                                                                                                                                                                                                                                                                                                                       |
|               |     Inner Join: CAST(store_sales.ss_quantity AS Int64) = taxi_trips.passenger_count                                                                                                                                                                                                                                                                                                                                                                                                                              |
|               |       BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |         Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
|               |  Projection: store_sales.ss_item_sk, store_sales.ss_quantity                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |   Filter: CAST(store_sales.ss_item_sk AS Int64) = Int64(180)                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |     TableScan: store_sales projection=[ss_item_sk, ss_quantity], partial_filters=[CAST(store_sales.ss_item_sk AS Int64) = Int64(180)]                                                                                                                                                                                                                                                                                                                                                                            |
|               |       BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |         Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
|               |  Projection: taxi_trips.passenger_count, taxi_trips.trip_distance_mi, taxi_trips.total_amount                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |   TableScan: taxi_trips projection=[passenger_count, trip_distance_mi, total_amount], full_filters=[taxi_trips.total_amount = Float64(3)]                                                                                                                                                                                                                                                                                                                                                                        |
```

Before 
```
Projection: store_sales.ss_sold_date_sk, store_sales.ss_sold_time_sk, store_sales.ss_item_sk, store_sales.ss_customer_sk, store_sales.ss_cdemo_sk, store_sales.ss_hdemo_sk, store_sales.ss_addr_sk, store_sales.ss_store_sk, store_sales.ss_promo_sk, store_sales.ss_ticket_number, store_sales.ss_quantity, store_sales.ss_wholesale_cost, store_sales.ss_list_price, store_sales.ss_sales_price, store_sales.ss_ext_discount_amt, store_sales.ss_ext_sales_price, store_sales.ss_ext_wholesale_cost, store_sales.ss_ext_list_price, store_sales.ss_ext_tax, store_sales.ss_coupon_amt, store_sales.ss_net_paid, store_sales.ss_net_paid_inc_tax, store_sales.ss_net_profit                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
|               |   Filter: CAST(store_sales.ss_item_sk AS Int64) = Int64(180)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |     TableScan: store_sales, partial_filters=[CAST(store_sales.ss_item_sk AS Int64) = Int64(180)]            
```

After: 
```
|               |  Projection: store_sales.ss_item_sk, store_sales.ss_quantity                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |   Filter: CAST(store_sales.ss_item_sk AS Int64) = Int64(180)                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |     TableScan: store_sales projection=[ss_item_sk, ss_quantity], partial_filters=[CAST(store_sales.ss_item_sk AS Int64) = Int64(180)]   
```

**Comment by goldmedal** [comment]
I'll review this PR tonight.

**Comment by goldmedal** [comment]
Thanks, @sgrebnov. I'm still reviewing. I wonder if it could be the default behavior for the `optimize_projections` rule instead of an option 🤔.

I tested some cases and found that the generated physical plans are the same whether this config is enabled.
```rust
    let mut config = ConfigOptions::new();
    config
        .optimizer
        .optimize_projections_preserve_existing_projections = true;
    let state = SessionConfig::from(config);
    let ctx = SessionContext::new_with_config(state);
    ctx.sql("CREATE TABLE test (a int, b int)").await.unwrap();
    let sql = "SELECT a, b FROM (SELECT a, b, count(*) from (SELECT a, b from test) group by a, b)";
    ctx.sql(sql).await.unwrap().explain(false, false).unwrap().show().await.unwrap();

    let mut config = ConfigOptions::new();
    config
        .optimizer
        .optimize_projections_preserve_existing_projections = false;
    let state = SessionConfig::from(config);
    let ctx = SessionContext::new_with_config(state);
    ctx.sql("CREATE TABLE test (a int, b int)").await.unwrap();
    let sql = "SELECT a, b FROM (SELECT a, b, count(*) from (SELECT a, b from test) group by a, b)";
    ctx.sql(sql).await.unwrap().explain(false, false).unwrap().show().await.unwrap();
```
The result is
```
Enable: optimize_projections_preserve_existing_projections
+---------------+--------------------------------------------------------------------------+
| plan_type     | plan                                                                     |
+---------------+--------------------------------------------------------------------------+
| logical_plan  | Projection: test.a, test.b                                               |
|               |   Aggregate: groupBy=[[test.a, test.b]], aggr=[[]]                       |
|               |     Projection: test.a, test.b                                           |
|               |       TableScan: test projection=[a, b]                                  |
| physical_plan | AggregateExec: mode=SinglePartitioned, gby=[a@0 as a, b@1 as b], aggr=[] |
|               |   MemoryExec: partitions=1, partition_sizes=[0]                          |
|               |                                                                          |
+---------------+--------------------------------------------------------------------------+
Disable optimize_projections_preserve_existing_projections
+---------------+--------------------------------------------------------------------------+
| plan_type     | plan                                                                     |
+---------------+--------------------------------------------------------------------------+
| logical_plan  | Aggregate: groupBy=[[test.a, test.b]], aggr=[[]]                         |
|               |   TableScan: test projection=[a, b]                                      |
| physical_plan | AggregateExec: mode=SinglePartitioned, gby=[a@0 as a, b@1 as b], aggr=[] |
|               |   MemoryExec: partitions=1, partition_sizes=[0]                          |
|               |                                                                          |
+---------------+--------------------------------------------------------------------------+
```

I'm not so familiar with the physical planner but I think if it doesn't cause too much effort for the physical planning, we can make it preserve the projection by default.

@alamb could you take a look? Thanks.

**Comment by alamb** [comment]
It seems confusing to disable only parts of an optimization rule for unparsing. I agree with @findepi  that it seems more logical to disable the entire rule rather than just some parts of it.

> Notice the additional projections added after joins.

What is the problem with the output? 

While there is an extra layer of projection it also doesn't seem like it would impact performance (it would likely be flattened by SQL optimizers)

Is the issue that it is hard to read for a human? If so, could you potentially write some sort of "projection pullup" pass that pulls up the parts you are interested in pulling up / removing?

**Comment by sgrebnov** [comment]
@alamb  - thank you for the feedback, the main challenge is that optimized plan can't be converted back to SQL after optimization (it is both added/moved and also removed projections). For example, the way how aggregation functions are unprojected rely on some assumptions about plan structure / etc.

The end goal is to get unnecessary columns pruned for the cases when plan needs to be split on sub-plans and executed separately. In this scenario pushing projections/columns downs ensure only required for final execution data is fetched. Keeping layout simple and human readable is nice but second.

Would implementing own optimizer using `optimize_projections` as a reference be a recommended approach in this case?

**Comment by findepi** [comment]
@sgrebnov is there a different example that could help understand the problem being solved?

the current PR description indicates the unparsed SQL being more complex and more nested that it otherwise could be, but why is this actually a problem? When splitting and routing queries, readability is nice to have, but not sure if it can be expected. Or there some other problem behind there?

**Comment by sgrebnov** [comment]
@findepi - sorry, my bad – I should have provided more context

We use Datafusion with [DataFusion Federation](https://github.com/spiceai/datafusion-federation/) to convert user queries into a LogicalPlan, then detect which parts of the plan belong to external execution engines. These parts are converted to SQL (unparsed) and executed by remote execution engines as part of the overall query execution. For example, in the scenario below, parts of the LogicalPlan are executed using external engines (MySQL and PostgreSQL) via unparsing corresponding sub-plans, with final aggregations or joins processed by DataFusion. All of this happens as part of DataFusion’s execution logic. If there are multiple external engines involved, only parts of the main plan are converted (see example below), so when we don’t have optimized/pushed-down projections, we end up fetching all columns. With projections optimization we propagate required columns to child nodes so only required columns could be fetched. Thus, the goal is to have projection columns pruning optimization enabled and to be able to unparse the logical plan back to SQL afterward. Please let me know if I should elaborate more on the challenges with the unparser after the optimization rule are applied.

```
                  ┌────────────────────────┐
                  │   Join / Aggregation   │               B and C are
                  └────────────────────────┘               available in an
                               ▲                           external database
                               │                           DBMS-2 (PostreSQL)
                               │
A is available in an           │                           Unparse -> SQL
external database in           ├─────────────────────┐
DBMS-1 (MySQL).                │┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─│─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
                  ┌────────────┘                     │
Unparse -> SQL    │             │                    │                      │
                  │                          ┌───────┴──────┐
                  │             │            │     Join     │               │
    ┌ ─ ─ ─ ─ ─ ─ ┼ ─ ─ ─ ─ ─ ┐              └───────▲──────┘
                  │             │                    │                      │
    │             │           │            ┌─────────┴──────────┐
         ┌────────┴───────┐     │          │                    │           │
    │    │     Scan A     │   │            │                    │
         └────────────────┘     │ ┌────────────────┐   ┌────────────────┐   │
    │                         │   │     Scan B     │   │     Scan C     │
                                │ └────────────────┘   └────────────────┘   │
    │                         │
     ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─  └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
```

**Comment by goldmedal** [comment]
I enabled `optimize_projections_preserve_existing_projections` to run the sqllogictests. I found that many projection push-downs do not work (the physical plan changes), and some queries fail.

### Push-down doesn’t work, but the query still succeeds (execution behavior may vary):
- Join
- Union
- Predicate
- Sort
- Window

### Query failures:
- `same_column_name_cross_join.slt`

It appears the physical planner struggles with additional projections. I believe `optimize_projections_preserve_existing_projections` is an unparser-specific config and shouldn't be used in query execution. However, I see the value in the additional projection for the unparser, as it helps identify the SQL body (`SELECT`) more easily.

As @alamb mentioned, having a config that changes part of an optimization can be confusing. Here are some potential solutions:
- Instead of modifying the existing optimization, create a separate `optimize_projection` rule for unparsing, allowing users to optimize their plan for SQL generation.
- Provide a config to switch `optimize_projection` into unparsing mode (though this might complicate the codebase).
- Improve the unparser to handle the original optimized plan.

**Comment by findepi** [comment]
> With projections optimization we propagate required columns to child nodes so only required columns could be fetched. 

makes sense!
this is important, agreed

can you reiterate or exemplify for the benefit of my understanding  what's the problem if you run all of `optimize_projection`?

**Comment by goldmedal** [comment]
> @goldmedal - does the "Push-down doesn’t work" part mean that projections are not pushed correctly (does not actually work) or something else? Could you please add more clarity on this. Going to test with `sqllogictests` as well - thank you for pointing to this.
> 

For example, the test in `joins.slt`.
https://github.com/apache/datafusion/blob/4e1f8391d4c2736ed92e5acba16dab51fb053aed/datafusion/sqllogictest/test_files/joins.slt#L1418-L1422
After enabling `optimize_projections_preserve_existing_projections`, the projection for `HashJoinExec` will be removed.

```diff
@@ -1425,10 +1433,9 @@ logical_plan
 01)Projection: count(alias1) AS count(DISTINCT join_t1.t1_id)
 02)--Aggregate: groupBy=[[]], aggr=[[count(alias1)]]
 03)----Aggregate: groupBy=[[join_t1.t1_id AS alias1]], aggr=[[]]
-04)------Projection: join_t1.t1_id
-05)--------Inner Join: join_t1.t1_id = join_t2.t2_id
-06)----------TableScan: join_t1 projection=[t1_id]
-07)----------TableScan: join_t2 projection=[t2_id]
+04)------Inner Join: join_t1.t1_id = join_t2.t2_id
+05)--------TableScan: join_t1 projection=[t1_id]
+06)--------TableScan: join_t2 projection=[t2_id]
 physical_plan
 01)ProjectionExec: expr=[count(alias1)@0 as count(DISTINCT join_t1.t1_id)]
 02)--AggregateExec: mode=Final, gby=[], aggr=[count(alias1)]
@@ -1436,7 +1443,7 @@ physical_plan
 04)------AggregateExec: mode=Partial, gby=[], aggr=[count(alias1)]
 05)--------AggregateExec: mode=SinglePartitioned, gby=[t1_id@0 as alias1], aggr=[]
 06)----------CoalesceBatchesExec: target_batch_size=2
-07)------------HashJoinExec: mode=Partitioned, join_type=Inner, on=[(t1_id@0, t2_id@0)], projection=[t1_id@0]
+07)------------HashJoinExec: mode=Partitioned, join_type=Inner, on=[(t1_id@0, t2_id@0)]
 08)--------------CoalesceBatchesExec: target_batch_size=2
 09)----------------RepartitionExec: partitioning=Hash([t1_id@0], 2), input_partitions=2
 10)------------------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
```
In this case, the pushdown doesn't work, but there are other similar cases. If you set this config to true by default, you can use the completed mode of the sqllogictests to find the cases in whichs the physical plan has changed.
```
cargo test --test sqllogictests -- --complete
```


> The main question is: do you think this work is very specific, meaning the improvement should be added directly to the project, or do you believe other developers could benefit from it, so we should continue exploring the best way to integrate/add this to the DataFusion?

Ideally, we should optimize the logical plans while retaining completeness, enabling us to unparse them back to SQL text easily—for example, by preserving `Projection`. Meanwhile, the physical planner can still generate an efficient physical plan with the necessary optimizations (e.g., Join with pushdown). If we can enable this configuration by default without altering any original physical plan, that would be the best-case scenario.

However, achieving this would be a huge epic, with many issues needing resolution before enabling it by default 🤔. 

One compromise is what I mentioned previously in https://github.com/apache/datafusion/pull/13267#issuecomment-2466215869. We could introduce a configuration (or simply rename `optimize_projections_preserve_existing_projections`) to switch the optimization to an unparse-friendly mode. Additionally, we should document that users should avoid executing SQL with this configuration enabled, as some SQL queries may fail in this mode.

https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/same_column_name_cross_join.slt

**Comment by sgrebnov** [comment]
@goldmedal  - thank you for detailed response. I'll review failing tests to get better understanding of why they are failing and if something can be improved within next few days and reply back. I like the idea of having this as a switch for unparser-friendly mode. As I've mentioned before there is similar case with predicates push down optimizer which could be controlled on the client side and I plan to investigate this more deeply with the target goal and priority of improving unparsing instead of adding something to the optimizer (but this still could be the case).

**Comment by findepi** [comment]
> Meanwhile, the physical planner can still generate an efficient physical plan with the necessary optimizations (e.g., Join with pushdown). If we can enable this configuration by default without altering any original physical plan, that would be the best-case scenario.

That would impose a limit on future optimizations on logical plans, which isn't desirable.

> 1. Column references require additional processing to project correctly

Can unparsing do this additional processing as needed for unparsing?

Generally, is unparsing expected to produce a query plan that is very similar to original query parsed by DF?
I hope that's a non goal, but if it is, it should really disable logical plan optimizations and maybe just do nothing more than projection pushdown.

**Comment by goldmedal** [comment]
> That would impose a limit on future optimizations on logical plans, which isn't desirable.
> 
@findepi I'm not sure about that. Why isn't it desirable? I aim to generate a logical plan with a suitable structure that's easy to convert to a physical plan and unparsed back to SQL. If we move some optimizations to the physical planner (e.g., removing unnecessary projection), we can still get the same execution plan without losing performance. 🤔

**Comment by berkaysynnada** [comment]
I didn’t have the chance to fully read through the discussion but IIUC, logical plans should always aim to provide the most optimal plan, independent of physical plans and their optimizations. We shouldn’t rely on physical plan behavior to ensure a truly modular engine. A user might reasonably expect the most optimal logical plan to be produced after logical optimizations and might proceed with custom physical rules that lack the expected capability.

**Comment by goldmedal** [comment]
> ..., logical plans should always aim to provide the most optimal plan, independent of physical plans and their optimizations. 
>

Thanks @berkaysynnada 
I see. If that's the main goal, I believe we should focus on continuously enhancing the unparsing behavior rather than modifying the optimization rule. This configuration should never be enabled by default. I suppose that's why @findepi mentioned it is desirable.

**Comment by findepi** [comment]
In other words, unparsing feature should not impose any constraints on logical plans produced by DF optimizer.

We should have a solid definition what is and what isn't a valid logical plan (and we should do a few changes in that area, https://github.com/apache/datafusion/issues/12723), but other than that unparsing should not impose any limitations on what logical plans DF creates.

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look 

I am trying to work down the review queue

**Comment by sgrebnov** [comment]
@alamb - makes sense, thank you. I'm going to close this PR as I don't think it makes sense to add this to DF, instead I'll create separate PR with just unparser improvement and will use custom optimizer rule.

/cc @findepi, @goldmedal

**Pull request closed without merging**